### PR TITLE
Refactor the ModelServer to let uvicorn handle multiple workers and use 'spawn' for mutiprocessing

### LINKED ISSUE #3406: Model server throws error when multiprocessing start method is already set
/kind bug

I’m trying to start a kserve model server with 2 workers like this
```model = ModelAPI(MODEL_NAME, MODEL_VERSION)
model.load()
kserve.ModelServer(workers=2).start([model])
```

But I get this error in the log
```
bin /app/venv/lib/python3.8/site-packages/bitsandbytes/libbitsandbytes_cpu.so
CUDA SETUP: Loading binary /app/venv/lib/python3.8/site-packages/bitsandbytes/libbitsandbytes_cpu.so...
{"asctime": "2024-02-01 16:51:03,775", "name": "torch.distributed.nn.jit.instantiator", "levelname": "INFO", "message": "Created a temporary directory at /tmp/tmp7h8azrb7"}
{"asctime": "2024-02-01 16:51:03,775", "name": "torch.distributed.nn.jit.instantiator", "levelname": "INFO", "message": "Writing /tmp/tmp7h8azrb7/_remote_module_non_scriptable.py"}
/app/venv/lib/python3.8/site-packages/bitsandbytes/cextension.py:33: UserWarning: The installed version of bitsandbytes was compiled without GPU support. 8-bit optimizers, 8-bit multiplication, and GPU quantization are unavailable.
  warn("The installed version of bitsandbytes was compiled without GPU support. "
{"asctime": "2024-02-01 16:52:11,342", "name": "root", "levelname": "INFO", "message": "PATH_TO_MODEL is set to ==>/mnt/models/"}
{"asctime": "2024-02-01 16:52:11,342", "name": "root", "levelname": "INFO", "message": "Registering model: llama-2-7b-chat-hf"}
{"asctime": "2024-02-01 16:52:11,343", "name": "root", "levelname": "INFO", "message": "Setting max asyncio worker threads as 8"}
{"asctime": "2024-02-01 16:52:11,344", "name": "root", "levelname": "INFO", "message": "Starting uvicorn with 2 workers"}
Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]
Loading checkpoint shards:  50%|█████     | 1/2 [00:48<00:48, 48.85s/it]
Loading checkpoint shards: 100%|██████████| 2/2 [01:06<00:00, 30.56s/it]
Loading checkpoint shards: 100%|██████████| 2/2 [01:06<00:00, 33.31s/it]
{"asctime": "2024-02-01 16:52:11,368", "name": "root", "levelname": "INFO", "message": "Starting gRPC server on [::]:8081"}
{"asctime": "2024-02-01 16:52:11,369", "name": "root", "levelname": "ERROR", "message": "uncaught exception", "exc_info": "Traceback (most recent call last):\n  File \"serve.py\", line 56, in <module>\n    kserve.ModelServer(workers=NUM_WORKERS).start([model])\n  File \"/app/venv/lib/python3.8/site-packages/kserve/model_server.py\", line 167, in start\n    asyncio.run(servers_task())\n  File \"/usr/lib/python3.8/asyncio/runners.py\", line 43, in run\n    return loop.run_until_complete(main)\n  File \"/usr/lib/python3.8/asyncio/base_events.py\", line 608, in run_until_complete\n    return future.result()\n  File \"/app/venv/lib/python3.8/site-packages/kserve/model_server.py\", line 165, in servers_task\n    await asyncio.gather(*servers)\n  File \"/app/venv/lib/python3.8/site-packages/kserve/model_server.py\", line 154, in serve\n    multiprocessing.set_start_method('fork')\n  File \"/usr/lib/python3.8/multiprocessing/context.py\", line 243, in set_start_method\n    raise RuntimeError('context has already been set')\nRuntimeError: context has already been set"}
Exception ignored in: <function Server.__del__ at 0x7f3b818b5ca0>
Traceback (most recent call last):
  File "/app/venv/lib/python3.8/site-packages/grpc/aio/_server.py", line 170, in __del__
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/common.pyx.pxi", line 118, in grpc._cython.cygrpc.schedule_coro_threadsafe
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/common.pyx.pxi", line 110, in grpc._cython.cygrpc.schedule_coro_threadsafe
  File "/usr/lib/python3.8/asyncio/base_events.py", line 425, in create_task
  File "/usr/lib/python3.8/asyncio/base_events.py", line 504, in _check_closed
RuntimeError: Event loop is closed
11:54
```

I would think kserve should handle multiprocessing gracefully since it offers passing more worker to the `kserve.ModelServer().start()` function but maybe I'm wrong?

[Slack Message](https://kubeflow.slack.com/archives/CH6E58LNP/p1706806945765699?thread_ts=1706806945.765699&cid=CH6E58LNP)

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

**What did you expect to happen:**

**What's the InferenceService yaml:**
[To help us debug please run `kubectl get isvc $name -n $namespace -oyaml` and paste the output]

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Istio Version:
- Knative Version:
- KServe Version: 0.10.0 - 0.12.0-rc1
- Kubeflow version:
- Cloud Environment:[k8s_istio/istio_dex/gcp_basic_auth/gcp_iap/aws/aws_cognito/ibm]
- Minikube/Kind version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

### LINKED ISSUE #3662: Cannot re-initialize CUDA in forked subprocess. To use CUDA with multiprocessing, you must use the 'spawn' start method
/kind bug

**What steps did you take and what happened:**

- **When I set the workers parameter to 2 or greater while cuda is active, it gives this error.**

` Cannot re-initialize CUDA in forked subprocess. To use CUDA with multiprocessing, you must use the 'spawn' start method`

- **If cuda is not active and the number of workers has a value of 2 or more, the model does not give prediction. timeout happens.**

**I am attaching 3 files as zipped. 
issue-open.py file can be used for reproducing the bug. 
Please run `python issue-open.py`
Please use demo.ipynb file for sending a request to model.** 

image

**Environment:**

- KServe Version: 0.12.1
[issue.zip](https://github.com/kserve/kserve/files/15193605/issue.zip)

### PR DESCRIPTION
**Pull request by sivanantha321**
**What this PR does / why we need it**:
- Refactored the ModelServer to let uvicorn handle multiple workers. This will remove the bottleneck of using 'fork' for multiprocessing. Now the **multiprocessing will use `spawn` method** which is compatible with torch cuda.

- Make FastAPI app instance easily accessible across the project so that users can **easily add middlewares** and custom exception handlers for custom models.
```python
from kserve.model_server import app

model = AlexNetModel(args.model_name)
model.load()
app.add_middleware(
        CORSMiddleware,
        allow_origins=["*"],
        allow_credentials=True,
        allow_methods=["*"],
        allow_headers=["*"],
)
ModelServer().start([model])
```

- ~Use uvloop eventpolicy~

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
- Fixes #3662
- Fixes #3406
- Closes #3407
- Closes #2954

**Type of changes**
Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
Refactor the ModelServer to let uvicorn handle multiple workers and use 'spawn' for mutiprocessing
```

**Re-running failed tests**

- `/rerun-all` - rerun all failed workflows.
- `/rerun-workflow <workflow name>` - rerun a specific failed workflow. Only one workflow name can be specified. Multiple /rerun-workflow commands are allowed per comment.

### CODE REVIEW

File: python/kserve/kserve/protocol/rest/server.py [hunk 1]
```diff
@@ -71,175 +58,60 @@ def install_signal_handlers(self) -> None:
...
+        rest_server = RESTServer(app, data_plane, model_repository_extension)
+        rest_server.create_application()
         app.add_middleware(
             TimingMiddleware,
             client=PrintTimings(),
             metric_namer=StarletteScopeToName(prefix="kserve.io", starlette_app=app),
         )
         self.cfg = uvicorn.Config(
-            app=app,
+            app="kserve.model_server:app",
```
Comment (sivanantha321) [inline python/kserve/kserve/protocol/rest/server.py:121]: This import string style is required for using `--workers` and `--reload` options. https://www.uvicorn.org/deployment/#running-programmatically

File: python/kserve/kserve/model_server.py [hunk 2]
```diff
@@ -360,12 +341,17 @@ def default_exception_handler(
         """Default exception handler for event loop.
 
         This is called when an exception occurs and no exception handler is set.
-        By default, this will shut down the server gracefully.
-
         This can be called by a custom exception handler that wants to defer to the default handler behavior.
         """
-        # gracefully shutdown the server
-        loop.run_until_complete(self.stop())
```
Comment (yuzisun) [inline python/kserve/kserve/model_server.py:384]: Do we still need to shut down the server?

File: python/kserve/kserve/model_server.py [hunk 3]
```diff
@@ -360,12 +341,17 @@ def default_exception_handler(
...
+        if "exception" in context:
+            logger.error(f"Caught exception: {context.get('exception')}")
+        logger.error(f"message: { context.get('message')}")
+        if "future" in context:
+            future = context["future"]
+            if future.done():
+                future_exception = future.exception()
+                if future_exception:
+                    logger.error(f"Future exception: {future_exception}")
```
Comment (yuzisun) [inline python/kserve/kserve/model_server.py]: what's the difference between exception and future_exception?

File: python/kserve/test/test_server.py [hunk 4]
```diff
@@ -769,12 +847,25 @@ def test_unload(self, http_server_client):
...
+        rest_server = RESTServer(
+            kserve_app, server.dataplane, server.model_repository_extension
+        )
+        rest_server.create_application()
+        yield server
+        kserve_app.routes.clear()
+
+    @pytest.fixture(scope="class")
+    def app(self, server):  # pylint: disable=no-self-use
+        mp = pytest.MonkeyPatch()
```
Comment (yuzisun) [inline python/kserve/test/test_server.py]: Why do we need monkey patch ?

File: python/kserve/test/test_inference_client.py [hunk 5]
```diff
@@ -26,21 +28,33 @@
...
         # model.load()  # Model not loaded, i.e. not ready
-        server = ModelServer()
         server.register_model(model)
         server.register_model(not_ready_model)
-        rest_server = RESTServer(server.dataplane, server.model_repository_extension)
-        app = rest_server.create_application()
+        yield kserve_app
+        await server.model_repository_extension.unload("TestModel")
+        await server.model_repository_extension.unload("NotReadyModel")
+        kserve_app.routes.clear()
```
Comment (yuzisun) [inline python/kserve/test/test_inference_client.py:48]: These are executed after app is terminated?

Comment (sivanantha321) [reply python/kserve/test/test_inference_client.py:48]: Yes, the statements after yield are used for cleanup

File: python/kserve/kserve/model_server.py (hunk 2 above)
Comment (sivanantha321) [reply python/kserve/kserve/model_server.py:384]: No Need. It will be handled by uvicorn.

File: python/kserve/kserve/model_server.py (hunk 3 above)
Comment (sivanantha321) [reply python/kserve/kserve/model_server.py]: It turns out both the exceptions are the same. Removed

File: python/kserve/test/test_server.py (hunk 4 above)
Comment (sivanantha321) [reply python/kserve/test/test_server.py]: Good catch. Actually I have tried a different approach before which needed this. This not needed now removing this.

**Review by yuzisun — APPROVED** [review]

**Comment by serdarildercaglar** [comment]
**Hello, I tested the update you made about workers. When we start the model with multiple workers, the model runs successfully. However, the model does not process requests in parallel as expected. First, it receives and processes one request, then when it is finished, it receives and processes the other request. For example, 5 requests of the same size were sent to the model simultaneously. Let the model process a single request in 20 seconds. Let the number of workers be 5 (our hardware resources are sufficient). Since the number of workers is 5, the model should process 5 requests at the same time simultaneously and return results. However since the model processes the requests sequentially, it takes 5X20 100 seconds to process all 5 requests.**
After running issue-open.py you can test the results using the test.ipynb file.

[files.zip](https://github.com/user-attachments/files/16553271/files.zip)

### Thank you for taking care of this issue.

**Pull request merged**